File: backend/app/schemas/visual.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


# 基础模型
class VisualToolBase(BaseModel):
    tool: str = Field(..., description="工具名称")
    name: str = Field(..., description="显示名称")
    description: Optional[str] = Field(None, description="工具描述")
    category_id: Optional[int] = Field(None, description="分类ID")


class VisualToolCreate(VisualToolBase):
    pass


class VisualToolUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    category_id: Optional[int] = None
    status: Optional[str] = None
    featured: Optional[bool] = None


class VisualToolResponse(VisualToolBase):
    id: int
    author_id: Optional[int] = None
    status: str
    featured: bool
    view_count: int
    like_count: int
    favorite_count: int
    comment_count: int
    usage_count: int
    created_at: datetime
    updated_at: datetime

    # 关联信息
    author_name: Optional[str] = None
    category_name: Optional[str] = None
    tags: List[str] = []

    # 用户交互状态
    is_liked: bool = False
    is_favorited: bool = False

    # 配置信息（从文件系统动态获取）
    params_schema: Optional[Dict[str, Any]] = None
    defaults: Optional[Dict[str, Any]] = None
    sample_data_filename: Optional[str] = None
    sample_image_url: Optional[str] = None
    docs_markdown: Optional[str] = None

    class Config:
        from_attributes = True


# 工具信息模型（文件系统）
class VisualToolInfo(BaseModel):
    tool: str = Field(..., description="工具名称，如 'line/basic'")
    name: str = Field(..., description="显示名称")
    description: str = Field(..., description="工具描述")
    category: str = Field(..., description="工具分类")
    tool_name: str = Field(..., description="工具名称（不包含分类）")
    params_schema: Dict[str, Any] = Field(default_factory=dict, description="参数模式")
    defaults: Dict[str, Any] = Field(default_factory=dict, description="默认参数")
    sample_data_filename: Optional[str] = Field(None, description="样本数据文件名")
    sample_image_url: Optional[str] = Field(None, description="样本图片URL")
    docs_markdown: Optional[str] = Field(None, description="文档Markdown")
    has_python: bool = Field(default=False, description="是否存在 Python 脚本")
    has_r: bool = Field(default=False, description="是否存在 R 脚本")
    has_js: bool = Field(default=True, description="是否存在 JS 版本")
    ggplot2: Optional[Dict[str, Any]] = Field(None, description="ggplot2 配置")
    heatmap: Optional[Dict[str, Any]] = Field(None, description="heatmap 配置")


# 工具分组模型
class VisualToolGroup(BaseModel):
    category: str = Field(..., description="分类名称")
    display_name: str = Field(..., description="分类显示名称")
    tools: List[VisualToolInfo] = Field(
        default_factory=list, description="该分类下的工具"
    )
    tool_count: int = Field(..., description="工具数量")


# 工具列表响应模型
class VisualToolsResponse(BaseModel):
    tools: List[VisualToolInfo] = Field(default_factory=list, description="所有工具")
    groups: List[VisualToolGroup] = Field(
        default_factory=list, description="按分类分组的工具"
    )
    total_tools: int = Field(..., description="总工具数")
    total_categories: int = Field(..., description="总分类数")
    category_stats: Dict[str, int] = Field(default_factory=dict, description="分类统计")


# 点赞/收藏模型
class UserToolLikeCreate(BaseModel):
    tool_id: int = Field(..., description="工具ID")


class UserToolFavoriteCreate(BaseModel):
    tool_id: int = Field(..., description="工具ID")


class UserToolLikeResponse(BaseModel):
    id: int
    user_id: int
    tool_id: int
    created_at: datetime

    class Config:
        from_attributes = True


class UserToolFavoriteResponse(BaseModel):
    id: int
    user_id: int
    tool_id: int
    created_at: datetime

    class Config:
        from_attributes = True


# 评论模型
class VisualToolCommentCreate(BaseModel):
    content: str = Field(..., description="评论内容", max_length=1000)
    parent_id: Optional[int] = Field(None, description="父评论ID")


class VisualToolCommentUpdate(BaseModel):
    content: Optional[str] = Field(None, description="评论内容", max_length=1000)
    status: Optional[str] = None


class VisualToolCommentResponse(BaseModel):
    id: int
    tool_id: int
    user_id: int
    content: str
    status: str
    like_count: int
    created_at: datetime
    updated_at: datetime

    # 关联信息
    author_name: Optional[str] = None
    author_avatar: Optional[str] = None
    parent_id: Optional[int] = None
    replies: List["VisualToolCommentResponse"] = []

    # 用户交互状态
    is_liked: bool = False

    class Config:
        from_attributes = True


# 评论点赞模型
class VisualToolCommentLikeCreate(BaseModel):
    comment_id: int = Field(..., description="评论ID")


class VisualToolCommentLikeResponse(BaseModel):
    id: int
    comment_id: int
    user_id: int
    created_at: datetime

    class Config:
        from_attributes = True


# 统计模型
class VisualToolStatsResponse(BaseModel):
    total_tools: int
    total_categories: int
    total_tags: int
    total_usage_count: int
    popular_tools: List[VisualToolResponse] = []
    recent_tools: List[VisualToolResponse] = []


# 绘图执行响应模型
class VisualRunResponse(BaseModel):
    """Response returned after running a visual tool."""

    success: bool
    message: Optional[str] = None
    output_files: List[str] = []  # 包含 PNG 和 PDF URL，第一个是 PNG，第二个是 PDF
    warnings: List[str] = []
    tool: Optional[str] = None
    used_params: Dict[str, Any] = Field(default_factory=dict)
    error_details: Optional[Dict[str, Any]] = Field(None, description="Detailed error information for recovery")
    data_info: Optional[Dict[str, Any]] = Field(None, description="Data format information")
    
    @property
    def image_url(self) -> Optional[str]:
        """Get PNG image URL"""
        return self.output_files[0] if len(self.output_files) > 0 else None
    
    @property
    def pdf_url(self) -> Optional[str]:
        """Get PDF URL"""
        return self.output_files[1] if len(self.output_files) > 1 else None
    
    @property
    def data_url(self) -> Optional[str]:
        """Get data URL if available"""
        return self.output_files[2] if len(self.output_files) > 2 else None


class VisualTaskResponse(BaseModel):
    """Response for async task submission."""

    task_id: str
    status: str
    message: str


class VisualTaskStatus(BaseModel):
    """Task status response."""

    task_id: str
    status: str
    progress: Optional[int] = None
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# 更新前向引用
VisualToolCommentResponse.model_rebuild()

File: frontend/src/libs/api/visual.ts
/**
 * Visual tools API functions
 */

import { apiRequest } from "./api";

// Visual tool types
export interface VisualToolInfo {
  tool: string; // 工具名称，如 'line/basic'
  name: string;
  display_name: string;
  description: string;
  category: string; // 添加分类信息
  tool_name: string; // 添加工具名称（不包含分类）
  params_schema: Record<string, any>;
  defaults: Record<string, any>;
  sample_data_filename?: string;
  sample_image_url?: string;
  docs_markdown?: string;
  has_python?: boolean; // 是否存在 Python 脚本
  has_r?: boolean; // 是否存在 R 脚本
  has_js?: boolean; // 是否存在 JS 版本
  ggplot2?: Record<string, any>; // ggplot2 配置
  heatmap?: Record<string, any>; // heatmap 配置
}

// 工具分组信息
export interface VisualToolGroup {
  category: string;
  display_name: string;
  tools: VisualToolInfo[];
  tool_count: number;
}

// 工具列表响应
export interface VisualToolsResponse {
  tools: VisualToolInfo[];
  groups: VisualToolGroup[];
  total_tools: number;
  total_categories: number;
  category_stats: Record<string, number>;
}

export interface VisualRunResponse {
  success: boolean;
  message: string;
  output_files?: string[]; // 包含 PNG 和 PDF URL，第一个是 PNG，第二个是 PDF
  tool: string;
  used_params: Record<string, any>;
}

export interface VisualRunParams {
  chart_type?: string;
  engine?: string;
  subdir?: string;
  data?: any[];
  [key: string]: any;
}

// 示例数据响应
export interface SampleDataResponse {
  success: boolean;
  data: any[] | Record<string, any[]>; // 可以是数组或字典
  data_type?: "list" | "dict"; // 数据类型标识
  message?: string;
}

// 文档响应
export interface DocsResponse {
  success: boolean;
  markdown: string;
  message?: string;
}

// Meta响应
export interface MetaResponse {
  success: boolean;
  meta: {
    name: string;
    description: string;
    params: Record<string, any>;
    additional_params?: Record<string, any>;
  };
  message?: string;
}

/**
 * Get all visual tools with grouping information
 */
export const getVisualTools = async (): Promise<VisualToolsResponse> => {
  try {
    const response = await apiRequest<VisualToolsResponse>("/visual/tools");
    return response;
  } catch (error) {
    console.error("Failed to fetch visual tools:", error);
    throw error;
  }
};

/**
 * Get visual tools grouped by category
 */
export const getVisualToolsGrouped = async (): Promise<VisualToolGroup[]> => {
  try {
    const response = await apiRequest<VisualToolGroup[]>(
      "/visual/tools/grouped"
    );
    return response;
  } catch (error) {
    console.error("Failed to fetch grouped visual tools:", error);
    throw error;
  }
};

/**
 * Get tools by category
 */
export const getVisualToolsByCategory = async (
  category: string
): Promise<VisualToolInfo[]> => {
  try {
    const response = await apiRequest<VisualToolInfo[]>(
      `/visual/tools/category/${category}`
    );
    return response;
  } catch (error) {
    console.error(`Failed to fetch tools for category ${category}:`, error);
    throw error;
  }
};

/**
 * Search visual tools
 */
export const searchVisualTools = async (
  query: string
): Promise<VisualToolInfo[]> => {
  try {
    const response = await apiRequest<VisualToolInfo[]>(
      `/visual/tools/search?q=${encodeURIComponent(query)}`
    );
    return response;
  } catch (error) {
    console.error(`Failed to search tools with query "${query}":`, error);
    throw error;
  }
};

/**
 * Get tool categories
 */
export const getVisualToolCategories = async (): Promise<string[]> => {
  try {
    const response = await apiRequest<string[]>("/visual/tools/categories");
    return response;
  } catch (error) {
    console.error("Failed to fetch tool categories:", error);
    throw error;
  }
};

/**
 * Get specific visual tool info by tool name (with underscore format)
 */
export const getVisualToolInfo = async (
  tool: string
): Promise<VisualToolInfo> => {
  try {
    const response = await apiRequest<VisualToolInfo>(`/visual/tools/${tool}`);
    return response;
  } catch (error) {
    console.error(`Failed to fetch visual tool info for ${tool}:`, error);
    throw error;
  }
};

/**
 * Run visual tool by tool name (with underscore format)
 */
export const runVisualTool = async (
  tool: string,
  params: VisualRunParams
): Promise<VisualRunResponse> => {
  try {
    const response = await apiRequest<VisualRunResponse>(
      `/visual/run/${tool}`,
      {
        method: "POST",
        data: params,
      }
    );
    return response;
  } catch (error) {
    console.error(`Failed to run visual tool ${tool}:`, error);
    throw error;
  }
};

/**
 * Run chart tool (new API)
 */
export const runChartTool = async (
  params: VisualRunParams
): Promise<VisualRunResponse> => {
  try {
    const response = await apiRequest<VisualRunResponse>("/visual/run-chart", {
      method: "POST",
      data: params,
    });
    return response;
  } catch (error) {
    console.error("Failed to run chart tool:", error);
    throw error;
  }
};

/**
 * Get sample data for a tool by tool name (with underscore format)
 */
export const getSampleData = async (
  tool: string
): Promise<SampleDataResponse> => {
  try {
    const response = await apiRequest<SampleDataResponse>(
      `/visual/tools/${tool}/sample-data`
    );
    return response;
  } catch (error) {
    console.error(`Failed to fetch sample data for ${tool}:`, error);
    throw error;
  }
};

/**
 * Get documentation for a tool by tool name (with underscore format)
 */
export const getToolDocs = async (tool: string): Promise<DocsResponse> => {
  try {
    const response = await apiRequest<DocsResponse>(
      `/visual/tools/${tool}/document`
    );
    return response;
  } catch (error) {
    console.error(`Failed to fetch docs for ${tool}:`, error);
    throw error;
  }
};

// 简化的图表生成接口
export interface SimpleChartResponse {
  success: boolean;
  message: string;
  image_url?: string;
  script_path?: string;
  data_path?: string;
}

export const generateSimpleChart = async (
  tool: string,
  parameters: { [key: string]: any },
  dataFile: File
): Promise<SimpleChartResponse> => {
  try {
    const formData = new FormData();
    formData.append("tool", tool);
    formData.append("parameters", JSON.stringify(parameters));
    formData.append("data_file", dataFile);

    const response = await fetch("/api/v1/visual-simple/generate-chart", {
      method: "POST",
      body: formData,
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    console.error(`Failed to generate chart for ${tool}:`, error);
    throw error;
  }
};

/**
 * Get meta.json for a tool by tool name (with underscore format)
 */
export const getToolMeta = async (tool: string): Promise<MetaResponse> => {
  try {
    const response = await apiRequest<MetaResponse>(
      `/visual/tools/${tool}/meta`
    );
    return response;
  } catch (error) {
    console.error(`Failed to fetch meta for ${tool}:`, error);
    throw error;
  }
};
